lib/block_view_message.go

```diff
@@ -825,6 +856,76 @@ func (bav *UtxoView) _connectMessagingGroup(
 		}
 	}
 
+	// V3 Messages: MUTING and UNMUTING members
+	if blockHeight >= bav.Params.ForkHeights.DeSoV3MessagesMutingBlockHeight {
+		if existingEntry != nil && !existingEntry.isDeleted {
+			value, operationTypeExists := txn.ExtraData["OperationType"]
+			if operationTypeExists {
+				// inline helper method
+				contains := func(memberList []*MessagingGroupMember, member *MessagingGroupMember) bool {
+					for _, a := range memberList {
+						if a == member {
+							return true
+						}
+					}
+					return false
+				}
+				// make deep copy of existingEntry.MuteList to prevent mempool errors
+				muteListCopy := make([]*MessagingGroupMember, len(existingEntry.MuteList))
+				copy(muteListCopy, existingEntry.MuteList)
+				// check if mute or unmute
+				if string(value) == MessagingGroupOperationMute {
+					for _, s := range txMeta.MessagingGroupMembers {
+						// Add s to muteList
+						// Make sure does not already exist to ensure no dups
+						if !contains(muteListCopy, s) {
+							muteListCopy = append(muteListCopy, s)
+						}
+					}
+				} else if string(value) == MessagingGroupOperationUnmute {
+					for _, s := range txMeta.MessagingGroupMembers {
+						for i, toUnmute := range muteListCopy {
+							if reflect.DeepEqual(toUnmute.GroupMemberPublicKey[:], s.GroupMemberPublicKey[:]) {
+								muteListCopy = append(muteListCopy[:i], muteListCopy[i+1:]...)
+								break
+							}
+						}
+					}
+				}
+				// Finally set the real existingEntry.MuteList to point to the clone
+				existingEntry.MuteList = muteListCopy
+				// Create a MessagingGroupEntry so we can add the entry to UtxoView.
+				messagingGroupEntry := MessagingGroupEntry{
+					GroupOwnerPublicKey:   &messagingGroupKey.OwnerPublicKey,
+					MessagingPublicKey:    messagingPublicKey,
+					MessagingGroupKeyName: NewGroupKeyName(txMeta.MessagingGroupKeyName),
+					MessagingGroupMembers: existingEntry.MessagingGroupMembers,
+					MuteList:              existingEntry.MuteList,
+					ExtraData:             existingEntry.ExtraData,
+				}
+				// Create a utxoOps entry, we make a copy of the existing entry.
+				var prevMessagingKeyEntry *MessagingGroupEntry
+				if existingEntry != nil && !existingEntry.isDeleted {
+					prevMessagingKeyEntry = &MessagingGroupEntry{}
+					rr := bytes.NewReader(EncodeToBytes(uint64(blockHeight), existingEntry))
+					if exists, err := DecodeFromBytes(prevMessagingKeyEntry, rr); !exists || err != nil {
+						return 0, 0, nil, errors.Wrapf(err,
+							"_connectMessagingGroup: Error decoding previous entry")
+					}
+				}
+				bav._setMessagingGroupKeyToMessagingGroupEntryMapping(&messagingGroupKey.OwnerPublicKey, &messagingGroupEntry)
+
+				// Construct UtxoOperation.
+				utxoOpsForTxn = append(utxoOpsForTxn, &UtxoOperation{
+					Type:                  OperationTypeMessagingKey,
+					PrevMessagingKeyEntry: prevMessagingKeyEntry,
+				})
+
+				return totalInput, totalOutput, utxoOpsForTxn, nil
```

Comment:
Hmm I don't love how this is duplicated, maybe we can somehow refactor the code so that we only create a utxoop once? (Nit: don't forget to merge ExtraData)

Reply:
fixing duplication with if else is causing high complexity and errors, so leaving it be for now

Reply:
if it ain't broke don't fix it

Reply:
ah, it would be reaaally nice if we could deduplicate this

Reply:
it was tedious but done :)